spec/android_send_app_size_metrics_spec.rb

```diff
@@ -0,0 +1,137 @@
+require_relative './spec_helper'
+
+describe Fastlane::Actions::AndroidSendAppSizeMetricsAction do
+  let(:test_data_dir) { File.join(File.dirname(__FILE__), 'test-data', 'app_size_metrics') }
+
+  def test_app_size_action(fake_aab_size:, fake_apks:, expected_payload:, **other_action_args)
+    in_tmp_dir do |tmp_dir|
+      # Arrange
+      output_file = File.join(tmp_dir, 'output-payload')
+      aab_path = File.join(tmp_dir, 'fake.aab')
+      File.write(aab_path, '-fake-aab-file-')
+      allow(File).to receive(:size).with(aab_path).and_return(fake_aab_size)
+
+      if other_action_args[:include_split_sizes] != false
+        # Arrange: fake that apkanalyzer exists
+        apkanalyzer_bin = File.join('__ANDROID_SDK_ROOT__FOR_TESTS__', 'cmdline-tools', 'latest', 'bin', 'apkanalyzer')
+        allow(described_class).to receive(:find_apkanalyzer_binary).and_return(apkanalyzer_bin)
+        allow(File).to receive(:executable?).with(apkanalyzer_bin).and_return(true)
+
+        # Arrange: fake that bundletool exists and mock its call to create fake apks with corresponding apkanalyzer calls mocks
+        allow(Fastlane::Action).to receive(:sh).with('command', '-v', 'bundletool', anything)
+        allow(Fastlane::Action).to receive(:sh).with('bundletool', 'build-apks', '--bundle', aab_path, '--output-format', 'DIRECTORY', '--output', anything) do |*args|
+          bundletool_tmpdir = args.last
+          FileUtils.mkdir(File.join(bundletool_tmpdir, 'splits'))
+          fake_apks.each do |apk_name, sizes|
+            apk_path = File.join(bundletool_tmpdir, 'splits', apk_name.to_s)
+            File.write(apk_path, "Fake APK file (#{sizes})")
+            allow(Fastlane::Action).to receive(:sh).with(apkanalyzer_bin, 'apk', 'file-size', apk_path, anything).and_return(sizes[0].to_s)
+            allow(Fastlane::Action).to receive(:sh).with(apkanalyzer_bin, 'apk', 'download-size', apk_path, anything).and_return(sizes[1].to_s)
+          end
+        end
+      end
+
+      # Act
+      code = run_described_fastlane_action(
+        api_url: File.join('file://localhost/', output_file),
+        aab_path: aab_path,
+        **other_action_args
+      )
+
+      # Asserts
+      expect(code).to eq(201)
+      expect(File).to exist(output_file)
+      gzip_disabled = other_action_args[:use_gzip_content_encoding] == false
+      generated_payload = gzip_disabled ? File.read(output_file) : Zlib::GzipReader.open(output_file, &:read)
+      # Compare the payloads as pretty-formatted JSON, to make the diff in test failures more readable if one happen
+      expect(JSON.pretty_generate(JSON.parse(generated_payload))).to eq(JSON.pretty_generate(expected_payload)), 'Decompressed JSON payload was not as expected'
+      # Compare the payloads as raw uncompressed data as a final check
+      expect(generated_payload).to eq(expected_payload.to_json)
```

Comment:
These two "triangulating" assertions (inappropriate because they're only two assertions, so not a triangle, but that's a term sometimes used for this pattern) are neat 😎